pkg/asset/cluster/metadata.go

```diff
@@ -0,0 +1,115 @@
+package cluster
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/aws"
+	"github.com/openshift/installer/pkg/asset/cluster/libvirt"
+	"github.com/openshift/installer/pkg/asset/cluster/openstack"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/types"
+	"github.com/pkg/errors"
+)
+
+const (
+	metadataFileName = "metadata.json"
+)
+
+// Metadata contains information needed to destroy clusters.
+type Metadata struct {
+	file *asset.File
+}
+
+var _ asset.WritableAsset = (*Metadata)(nil)
+
+// Name returns the human-friendly name of the asset.
+func (m *Metadata) Name() string {
+	return "Metadata"
+}
+
+// Dependencies returns the direct dependencies for the metadata
+// asset.
+func (m *Metadata) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&installconfig.ClusterID{},
+		&installconfig.InstallConfig{},
+	}
+}
+
+// Generate generates the metadata asset.
+func (m *Metadata) Generate(parents asset.Parents) (err error) {
+	clusterID := &installconfig.ClusterID{}
+	installConfig := &installconfig.InstallConfig{}
+	parents.Get(clusterID, installConfig)
+
+	if installConfig.Config.Platform.None != nil {
+		return errors.New("metadata cannot be created with platform set to 'none'")
+	}
+
+	metadata := &types.ClusterMetadata{
+		ClusterName: installConfig.Config.ObjectMeta.Name,
+		ClusterID:   clusterID.ClusterID,
+	}
+
+	switch {
+	case installConfig.Config.Platform.AWS != nil:
+		metadata.ClusterPlatformMetadata.AWS = aws.Metadata(clusterID.ClusterID, installConfig.Config)
+	case installConfig.Config.Platform.Libvirt != nil:
+		metadata.ClusterPlatformMetadata.Libvirt = libvirt.Metadata(installConfig.Config)
+	case installConfig.Config.Platform.OpenStack != nil:
+		metadata.ClusterPlatformMetadata.OpenStack = openstack.Metadata(clusterID.ClusterID, installConfig.Config)
+	default:
+		return errors.Errorf("no known platform")
+	}
+
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return errors.Wrap(err, "failed to Marshal ClusterMetadata")
+	}
+
+	m.file = &asset.File{
+		Filename: metadataFileName,
+		Data:     data,
+	}
+
+	return nil
+}
+
+// Files returns the FileList generated by the asset.
+func (m *Metadata) Files() []*asset.File {
+	return []*asset.File{m.file}
+}
+
+// Load returns error if the metadata is already on-disk, because we want to
+// prevent user from accidentally re-launching the cluster.
+func (m *Metadata) Load(f asset.FileFetcher) (found bool, err error) {
+	_, err = f.FetchByName(metadataFileName)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, errors.Errorf("%q already exists.  There may already be a running cluster", metadataFileName)
```

Comment:
Won't this prevent running the `cluster` target after the `ignition-configs` target?

Reply:
> Won't this prevent running the cluster target after the `ignition-configs` target?

Hmm, maybe.  Is there an easy way to allow loading and only complain if it has changed?

Reply:
My suggestion is that, for now, `Load` should return `false` always. The installer will ignore any changes to `metadata.json`. In the future, perhaps we should introduce a read-only asset that would cause the installer to warn (or fail) in the face of changes.

Reply:
> My suggestion is that, for now, `Load` should return `false` always.

Done with 76e67a4 -> a9afdc9.